hypershift-operator/controllers/awsendpointservice/awsendpointservice_controller.go

```diff
@@ -0,0 +1,208 @@
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	awsutil "github.com/openshift/hypershift/cmd/infra/aws/util"
+	"github.com/openshift/hypershift/support/upsert"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	finalizer                              = "hypershift.openshift.io/finalizer"
+	endpointServiceDeletionRequeueDuration = time.Duration(5 * time.Second)
+)
+
+type AWSEndpointServiceReconciler struct {
+	client.Client
+	upsert.CreateOrUpdateProvider
+	Region      string
+	ec2Client   ec2iface.EC2API
+	elbv2Client elbv2iface.ELBV2API
+}
+
+func (r *AWSEndpointServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Only add this controller to the manager if we have a non-empty credentials file
+	credsFile := os.Getenv("AWS_SHARED_CREDENTIALS_FILE")
+	if len(credsFile) == 0 {
+		return nil
+	}
+	credsBytes, err := ioutil.ReadFile(credsFile)
+	if err != nil || len(credsBytes) == 0 {
+		return nil
+	}
+
+	_, err = ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.AWSEndpointService{}).
+		Build(r)
+	if err != nil {
+		return fmt.Errorf("failed setting up with a controller manager: %w", err)
+	}
+
+	awsSession := awsutil.NewSession("hypershift-operator")
+	awsConfig := awsutil.NewConfig(credsFile, r.Region)
+	r.ec2Client = ec2.New(awsSession, awsConfig)
+	r.elbv2Client = elbv2.New(awsSession, awsConfig)
+
+	return nil
+}
+
+func (r *AWSEndpointServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// fetch the AWSEndpointService
+	awsEndpointService := &hyperv1.AWSEndpointService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+	err := r.Get(ctx, client.ObjectKeyFromObject(awsEndpointService), awsEndpointService)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get resource: %w", err)
+	}
+
+	// Return early if deleted
+	if !awsEndpointService.DeletionTimestamp.IsZero() {
+		completed, err := r.delete(ctx, awsEndpointService)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to delete resource: %w", err)
+		}
+		if !completed {
+			return ctrl.Result{RequeueAfter: endpointServiceDeletionRequeueDuration}, nil
+		}
+		if controllerutil.ContainsFinalizer(awsEndpointService, finalizer) {
+			controllerutil.RemoveFinalizer(awsEndpointService, finalizer)
+			if err := r.Update(ctx, awsEndpointService); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Ensure the awsEndpointService has a finalizer for cleanup
+	if !controllerutil.ContainsFinalizer(awsEndpointService, finalizer) {
+		controllerutil.AddFinalizer(awsEndpointService, finalizer)
+		if err := r.Update(ctx, awsEndpointService); err != nil {
+			if apierrors.IsConflict(err) {
```

Comment:
returning the err will requeue any way, why do we need to make this distinction?

Reply:
because one results in an error in the log and the other is just bad timing